go-controller/pkg/ovn/egressfirewall.go

```diff
@@ -0,0 +1,331 @@
+
+	egressfirewallapi "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/egressfirewall/v1"
+
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog"
+)
+
+const (
+	defaultStartPriority = 2000
+)
+
+type egressFirewall struct {
+	name        string
+	namespace   string
+	egressRules []*egressFirewallRule
+}
+
+type egressFirewallRule struct {
+	id     int
+	access egressfirewallapi.EgressFirewallRuleType
+	ports  []port
+	to     destination
+}
+
+type port struct {
+	protocol string
+	port     int32
+}
+
+type destination struct {
+	cidrSelector string
+}
+
+func newEgressFirewall(egressFirewallPolicy *egressfirewallapi.EgressFirewall) *egressFirewall {
+	ef := &egressFirewall{
+		name:        egressFirewallPolicy.Name,
+		namespace:   egressFirewallPolicy.Namespace,
+		egressRules: make([]*egressFirewallRule, 0),
+	}
+	return ef
+}
+
+func newEgressFirewallRule(rawEgressFirewallRule egressfirewallapi.EgressFirewallRule, id int) (*egressFirewallRule, error) {
+	efr := &egressFirewallRule{
+		id:     id,
+		access: rawEgressFirewallRule.Type,
+	}
+
+	_, _, err := net.ParseCIDR(rawEgressFirewallRule.To.CIDRSelector)
+	if err != nil {
+		return nil, err
+
+	}
+	efr.to.cidrSelector = rawEgressFirewallRule.To.CIDRSelector
+
+	for _, rawPort := range rawEgressFirewallRule.Ports {
+		port := port{}
+		port.protocol = rawPort.Protocol
+		port.port = rawPort.Port
+		efr.ports = append(efr.ports, port)
+	}
+
+	return efr, nil
+}
+
+func (oc *Controller) addEgressFirewall(egressFirewall *egressfirewallapi.EgressFirewall) {
+	klog.Infof("Adding egress Firewall %s in namespace %s", egressFirewall.Name, egressFirewall.Namespace)
+	nsInfo, err := oc.waitForNamespaceLocked(egressFirewall.Namespace)
+	if err != nil {
+		klog.Errorf("Failed to wait for namespace %s event (%v)",
+			egressFirewall.Namespace, err)
+		return
+	}
+	defer nsInfo.Unlock()
+
+	if nsInfo.egressFirewallPolicy != nil {
+		klog.Errorf("Attempting to add egressFirewall %s to namespace %s when it already has an egressFirewall",
+			egressFirewall.Name, egressFirewall.Namespace)
+		return
+	}
+
+	ef := newEgressFirewall(egressFirewall)
+	nsInfo.egressFirewallPolicy = ef
+	for i, egressFirewallRule := range egressFirewall.Spec.Rules {
+		//process Rules into egressFirewallRules for egressFirewall struct
+		efr, err := newEgressFirewallRule(egressFirewallRule, i)
+		if err != nil {
+			klog.Errorf("Cannot create EgressFirewall Rule for destination %s to namespace %s - %v",
+				egressFirewallRule.To.CIDRSelector, egressFirewall.Namespace, err)
+			continue
+
+		}
+		ef.egressRules = append(ef.egressRules, efr)
+	}
+
+	existingNodes, err := oc.kube.GetNodes()
+	if err != nil {
+		klog.Errorf("Unable to add egressfirewall %s, cannot list nodes: %s", egressFirewall.Name, err)
+	}
+	for _, node := range existingNodes.Items {
+		joinSwitch := joinSwitch(node.Name)
+		err = ef.addACLToJoinSwitch(joinSwitch, nsInfo.addressSet.GetIPv4HashName(), nsInfo.addressSet.GetIPv6HashName())
```

Comment:
@trozet though, if for some reason a node gets removed during that time would the mega-nbctl command fail because the join switch wasn't present?

Reply:
well I think that is another reason we need caching for the nodes. If the node is being deleted, the cache should be locked, and then we would never add the acl to the now deleted join switch (because we would be waiting for a cache lock in addACLToJoinSwitches). Regardless of configuring the switches with individual commands or batching them, I still think you can call the function one time, because the rest of the code in that function is the same for every switch.

With the move to a single join switch, the debate of batching the commands or not becomes less of an issue IMO the cost savings of batching > the window of potential race between node deletion and adding the ACLs.

Reply:
@trozet I don't think we can batch commands at this point. Perhaps I wasn't clear, but my scenario was:

1) addEgressFirewall() starts building single-transaction ovn-nbctl command
2) WatchNodes() gets and processes a delete event for a node and removes the node switch
3) addEgressFirewall() finishes building ovn-nbctl command and executes
4) entire ovn-nbctl batch transaction fails because one logical switch doesn't exist
5) doesn't get retried until the egressfireall CR is changed

I'd rather just do individual ones for now, and mega-batch with a retry later or something. These object shouldn't be changing frequently.

Reply:
@dcbw I understood you, I just don't agree. But it's fine because we will be going to single join switch, so in this patch I don't mind doing individual commands. However, my comment still stands I think to reduce to a single function call and just pass a slice of join switches, and not re-run the join switch agnostic ACL composing N number of node times.